pyam/core.py

```diff
@@ -733,118 +727,140 @@ def check_aggregate(self, variable, components=None, units=None,
 
-            diff = pd.concat([diff], keys=[variable], names=['variable'])
+            return IamDataFrame(diff, variable=variable).timeseries()
+
+    def aggregate_region(self, variable, region='World', subregions=None,
+                         components=None, unit=None, append=False):
+        """Compute the aggregate of timeseries over a number of regions
+        including variable components only defined at the `region` level
 
-            return diff.unstack().rename_axis(None, axis=1)
+        Parameters
+        ----------
+        variable: str
+            variable for which the aggregate should be computed
+        region: str, default 'World'
+            dimension
+        subregions: list of str
+            list of subregions, defaults to all regions other than `region`
+        components: list of str
+            list of variables, defaults to all sub-categories of `variable`
+            included in `region` but not in any of `subregions`
+        unit: str or list of str
+            filter variable and variable components for given unit(s)
+        append: bool, default False
+            append the aggregate timeseries to `data` and return None,
+            else return aggregate timeseries
+        """
+        # default subregions to all regions other than `region`
+        if subregions is None:
+            rows = self._apply_filters(variable=variable)
+            subregions = set(self.data[rows].region) - set([region])
 
-    def check_aggregate_regions(self, variable, region='World',
-                                components=None, units=None,
-                                exclude_on_fail=False, **kwargs):
+        if not len(subregions):
+            msg = 'cannot aggregate variable `{}` to `{}` because it does not'\
+                  ' exist in any subregion'
+            logger().info(msg.format(variable, region))
+
+            return
+
+        # compute aggregate over all subregions
+        subregion_df = self.filter(region=subregions, unit=unit)
+        cols = ['region', 'variable']
+        _data = _aggregate(subregion_df.filter(variable=variable).data, cols)
+
+        # add components at the `region` level, defaults to all variables one
+        # level below `variable` that are only present in `region`
+        region_df = self.filter(region=region, unit=unit)
+        components = components or (
+            set(region_df._variable_components(variable)).difference(
+                subregion_df._variable_components(variable)))
+
+        if len(components):
+            rows = region_df._apply_filters(variable=components)
+            _data = _data.add(_aggregate(region_df.data[rows], cols),
+                              fill_value=0)
+
+        if append is True:
+            self.append(_data, region=region, variable=variable, inplace=True)
+        else:
+            return _data
+
+    def check_aggregate_region(self, variable, region='World', subregions=None,
```

Comment:
here I would actually make the same comment on the `mapping` arg, let me know if you disagree. @znicholls may also have an opinion here..

Reply:
See response above.